### DISCUSSION TITLE
api route example on nextjs-prisma-postgres guide

### DISCUSSION THREAD

**Discussion by viktor992** [prisma/prisma#19400] [Q&A] [answered]
I ran into an issue when using nextAuth in an API route as recommended in the "[How to Build a Fullstack App with Next.js, Prisma, and Vercel Postgres](https://vercel.com/guides/nextjs-prisma-postgres)" guide on the Vercel website. Since I couldn't find the repository for this guide, I'm initiating a discussion here to see if it's possible to resolve it.

[Here's](https://vercel.com/guides/nextjs-prisma-postgres#step-8.-add-new-post-functionality) how the guide currently recommends implementing an api route using next-auth:

```ts
// pages/api/post/index.ts

import { getSession } from 'next-auth/react';
import prisma from '../../../lib/prisma';

// POST /api/post
// Required fields in body: title
// Optional fields in body: content
export default async function handle(req, res) {
  const { title, content } = req.body;
  const session = await getSession({ req });
  const result = await prisma.post.create({
    data: {
      title: title,
      content: content,
      author: { connect: { email: session?.user?.email } },
    },
  });
  res.json(result);
}
```

However, every time I call this endpoint from the client side, the `getSession` method returns `null` and the `CLIENT_FETCH_ERROR` is shown in the console.

I reviewed the next-auth docs and found that they [recommend](https://next-auth.js.org/tutorials/securing-pages-and-api-routes#using-getserversession) using the `getServerSession()` method instead. So, after making the following changes the api route worked as intended.

First, I exported the nextauth options from `pages/api/auth/[...nextauth].ts`

```ts
// pages/api/auth/[...nextauth].ts

import { NextApiHandler } from "next";
import NextAuth from "next-auth";
import { PrismaAdapter } from "@next-auth/prisma-adapter";
import GitHubProvider from "next-auth/providers/github";
import prisma from "../../../lib/prisma";

export const options = { // <--- exported the nextauth options
  providers: [
    GitHubProvider({
      clientId: process.env.GITHUB_ID,
      clientSecret: process.env.GITHUB_SECRET,
    }),
  ],
  adapter: PrismaAdapter(prisma),
  secret: process.env.SECRET,
};

const authHandler: NextApiHandler = (req, res) => NextAuth(req, res, options);
export default authHandler;

```

Then, I modified the pages/api/post/index.ts file to import the nextauth options and use the `getServerSession` method.

```ts
// pages/api/post/index.ts
// import { getSession } from 'next-auth/react'; // <--- removed getSession import
import { getServerSession } from "next-auth/next" // <--- imported getServerSession from "next-auth/next"
import { options as authOptions } from "../auth/[...nextauth]" // <--- imported authOptions
import prisma from "../../../lib/prisma";

// POST /api/post
// Required fields in body: title
// Optional fields in body: content
export default async function handle(req, res) {
  const { title, content } = req.body;

  // const session = await getSession({ req }); // <--- removed getSession call
  const session = await getServerSession(req, res, authOptions); // <--- used the getServerSession instead
  const result = await prisma.post.create({
    data: {
      title: title,
      content: content,
      author: { connect: { email: session?.user?.email } },
    },
  });
  res.json(result);
}

```

This was the only way I managed to make the application work, and if you agree, we could modify the guide to include these changes to prevent others from experiencing the same issue. I look forward to your feedback and comments.

**Comment by nurul3101** [comment, marked as answer]
Hello @viktor992 👋 

This was discussed recently in this GitHub Discussion and we do plan to upgrade that guide soon.

- https://github.com/prisma/prisma/discussions/19276

Can you have a look at this discussion?

**Reply by viktor992** [reply]
Oh, it's exactly the same issue. I'm sorry, I hadn't come across that discussion. I'll add a reference to the solution I proposed here in case it's helpful to someone. Thanks @nurul3101!

**Comment by kimoycatayas** [comment]
I tried this solution today and It works. Thanks @viktor992

**Reply by nurul3101** [reply]
Thanks for confirming! 😄

**Reply by kimoycatayas** [reply]
Thank you for the support @nurul3101

**Comment by gloriliale** [comment]
Thank you for posting this solution, @viktor992, it was really helpful for completing my project!

@nurul3101 , I worked through the same guide ([How to Build a Fullstack App with Next.js, Prisma, and Vercel Postgres](https://vercel.com/guides/nextjs-prisma-postgres)) and documented a few things that need to be updated for it to run successfully. Let me know if that's of use to you and if so, where to send it. Thanks!

**Reply by georgernstgraf** [reply]
Same for me here, thanks @viktor992. Interrestingly the tutorial is still out of date though.

**Comment by ChristopherJH** [comment]
Thanks for the solution! I was stuck on the same problem. Weird they still haven't updated the guide?

**Reply by georgernstgraf** [reply]
Yes that's interesting. They seem to have more focus on creating than on maintenance... Glad my solution helped!

**Reply by georgernstgraf** [reply]
Nope it was not my solution :/
